### DISCUSSION TITLE
TabView subviews redundant init

### DISCUSSION THREAD

**Discussion by rmncv** [pointfreeco/swift-composable-architecture#347] [Q&A] [answered]
Hello :)
Well, I’m wondering maybe you could navigate me or give some advice about what to do with this.

I have a root view (TabView), with AppState’s store, and each of the tabs is the scope of state to its local one. But, whenever I’m receiving any action, the whole body of the root view is recreated, and, as a result, each view creates one more time. How to deal with this?

I added debug to my app reducer, and on first tab switch, I see something like this:

received action:
  AppAction.shoppingList(
    ShoppingListFeatureAction.onAppear
  )
  (No state changes)

✅ FridgesFeatureView init
✅ ShoppingListFeatureView init
✅ RecipesListView init

And this is why I’m catching exceptions:

Thread 1: Fatal error: The store was sent the action AppAction.recipesTab(.onAppear) while it was already 
processing another action.

This can happen for a few reasons:

* The store was sent an action recursively. This can occur when you run an effect directly in the reducer, rather than returning it from the reducer. Check the stack (⌘7) to find frames corresponding to one of your reducers. That code should be refactored to not invoke the effect directly.

* The store has been sent actions from multiple threads.
![Xcode_2021-01-15 18-24-28@2x](https://user-images.githubusercontent.com/8841444/104756678-c77aec00-5764-11eb-9e2c-220a9aae3b9d.png)
![Xcode_2021-01-15 18-24-39@2x](https://user-images.githubusercontent.com/8841444/104756692-c9dd4600-5764-11eb-9f48-4303eb4622e0.png)
![Xcode_2021-01-15 18-24-54@2x](https://user-images.githubusercontent.com/8841444/104756696-cb0e7300-5764-11eb-9fc9-eb4f3dd1daee.png)

**Comment by mbrandonw** [comment, marked as answer]
If you are seeing too many views created/computed then it probably means your `WithViewStore` could be focused on a smaller set of state. For example, in the screenshot you posted you have `WithViewStore(self.store)`, but then it seems that you aren't actually accessing any state from `viewStore`. That means the view will be recomputed any time actions are sent even though the state is never actually used to compute the body of the view.

So what you can do is `.scope` your store so that it holds only the pieces of state you care about. In this case you don't actually care about any state, and so you can use `WithViewStore(self.store.stateless)`.
